File: examples/flutter/lib/ren_sdk.dart
/// Ren SDK для Flutter
///
/// Dart FFI обёртка для Ren SDK
library ren_sdk;

import 'dart:ffi';
import 'dart:io';
import 'dart:convert';
import 'package:ffi/ffi.dart';

/// Результат операции (C структура)
class RenResult extends Struct {
  @Int32()
  external int code;

  external Pointer<Utf8>? message;

  bool get isSuccess => code == 0;

  void throwIfError() {
    if (!isSuccess) {
      final msg = message?.toDartString();
      final freeFunc = _dylib
          .lookupFunction<
            Void Function(Pointer<Utf8>),
            void Function(Pointer<Utf8>)
          >('ren_sdk_free_string');
      if (message != null) {
        freeFunc(message!);
      }
      throw RenSDKError(msg ?? 'Unknown error', code);
    }
  }

  static DynamicLibrary? _dylib;
  static DynamicLibrary get _library {
    _dylib ??= RenSDK._loadLibrary();
    return _dylib!;
  }
}

/// Ошибка SDK
class RenSDKError implements Exception {
  final String message;
  final int code;

  RenSDKError(this.message, [this.code = 0]);

  @override
  String toString() => 'RenSDKError: $message (code: $code)';
}

/// Результат операции (C структура)
class RenResult extends Struct {
  @Int32()
  external int code;

  external Pointer<Utf8>? message;

  bool get isSuccess => code == 0;

  void throwIfError() {
    if (!isSuccess) {
      final msg = message?.toDartString();
      throw RenSDKError(msg ?? 'Unknown error', code);
    }
  }
}

/// Клиент Ren SDK для Flutter
class RenSDK {
  final DynamicLibrary _dylib;
  final Pointer<Void> _handle;

  RenSDK._(this._dylib, this._handle);

  /// Создаёт новый клиент SDK
  factory RenSDK.create(String baseUrl) {
    final dylib = _loadLibrary();
    final createFunc = dylib
        .lookupFunction<
          Pointer<Void> Function(Pointer<Utf8>),
          Pointer<Void> Function(Pointer<Utf8>)
        >('ren_sdk_client_new');

    final baseUrlPtr = baseUrl.toNativeUtf8();
    final handle = createFunc(baseUrlPtr);
    malloc.free(baseUrlPtr);

    if (handle == nullptr) {
      throw RenSDKError('Failed to create SDK client');
    }

    return RenSDK._(dylib, handle);
  }

  /// Загружает нативную библиотеку
  static DynamicLibrary _loadLibrary() {
    if (Platform.isWindows) {
      return DynamicLibrary.open('ren_sdk.dll');
    } else if (Platform.isLinux) {
      return DynamicLibrary.open('libren_sdk.so');
    } else if (Platform.isMacOS) {
      return DynamicLibrary.open('libren_sdk.dylib');
    } else if (Platform.isAndroid) {
      return DynamicLibrary.open('libren_sdk.so');
    } else if (Platform.isIOS) {
      return DynamicLibrary.open('libren_sdk.a');
    } else {
      throw UnsupportedError('Platform not supported');
    }
  }

  /// Устанавливает токен авторизации
  void setToken(String token) {
    final setTokenFunc = _dylib
        .lookupFunction<
          RenResult Function(Pointer<Void>, Pointer<Utf8>),
          RenResult Function(Pointer<Void>, Pointer<Utf8>)
        >('ren_sdk_client_set_token');

    final tokenPtr = token.toNativeUtf8();
    final result = setTokenFunc(_handle, tokenPtr);
    malloc.free(tokenPtr);

    result.throwIfError();
  }

  /// Получает токен
  String? getToken() {
    final getTokenFunc = _dylib
        .lookupFunction<
          Pointer<Utf8> Function(Pointer<Void>),
          Pointer<Utf8> Function(Pointer<Void>)
        >('ren_sdk_client_get_token');

    final tokenPtr = getTokenFunc(_handle);
    if (tokenPtr == nullptr) {
      return null;
    }

    final token = tokenPtr.toDartString();
    final freeFunc = _dylib
        .lookupFunction<
          Void Function(Pointer<Utf8>),
          void Function(Pointer<Utf8>)
        >('ren_sdk_free_string');
    freeFunc(tokenPtr);

    return token;
  }

  /// Выполняет вход в систему
  void login(String login, String password, {bool rememberMe = false}) {
    final loginFunc = _dylib
        .lookupFunction<
          RenResult Function(
            Pointer<Void>,
            Pointer<Utf8>,
            Pointer<Utf8>,
            Int32,
          ),
          RenResult Function(Pointer<Void>, Pointer<Utf8>, Pointer<Utf8>, int)
        >('ren_sdk_login');

    final loginPtr = login.toNativeUtf8();
    final passwordPtr = password.toNativeUtf8();
    final result = loginFunc(
      _handle,
      loginPtr,
      passwordPtr,
      rememberMe ? 1 : 0,
    );
    malloc.free(loginPtr);
    malloc.free(passwordPtr);

    result.throwIfError();
  }

  /// Получает профиль текущего пользователя
  Map<String, dynamic> getMe() {
    final getMeFunc = _dylib
        .lookupFunction<
          Pointer<Utf8> Function(Pointer<Void>),
          Pointer<Utf8> Function(Pointer<Void>)
        >('ren_sdk_get_me');

    final jsonPtr = getMeFunc(_handle);
    if (jsonPtr == nullptr) {
      throw RenSDKError('Failed to get profile');
    }

    final jsonStr = jsonPtr.toDartString();
    final freeFunc = _dylib
        .lookupFunction<
          Void Function(Pointer<Utf8>),
          void Function(Pointer<Utf8>)
        >('ren_sdk_free_string');
    freeFunc(jsonPtr);

    return jsonDecode(jsonStr) as Map<String, dynamic>;
  }

  /// Получает список чатов
  List<Map<String, dynamic>> getChats() {
    final getChatsFunc = _dylib
        .lookupFunction<
          Pointer<Utf8> Function(Pointer<Void>),
          Pointer<Utf8> Function(Pointer<Void>)
        >('ren_sdk_get_chats');

    final jsonPtr = getChatsFunc(_handle);
    if (jsonPtr == nullptr) {
      throw RenSDKError('Failed to get chats');
    }

    final jsonStr = jsonPtr.toDartString();
    final freeFunc = _dylib
        .lookupFunction<
          Void Function(Pointer<Utf8>),
          void Function(Pointer<Utf8>)
        >('ren_sdk_free_string');
    freeFunc(jsonPtr);

    return (jsonDecode(jsonStr) as List).cast<Map<String, dynamic>>();
  }

  /// Освобождает ресурсы
  void dispose() {
    final freeFunc = _dylib
        .lookupFunction<
          Void Function(Pointer<Void>),
          void Function(Pointer<Void>)
        >('ren_sdk_client_free');
    freeFunc(_handle);
  }
}

/// Статические функции для криптографии
class RenSDKCrypto {
  static DynamicLibrary? _dylib;

  static DynamicLibrary get _library {
    _dylib ??= RenSDK._loadLibrary();
    return _dylib!;
  }

  /// Генерирует пару ключей
  static Map<String, String> generateKeypair() {
    final func = _library
        .lookupFunction<Pointer<Utf8> Function(), Pointer<Utf8> Function()>(
          'ren_sdk_generate_keypair',
        );

    final jsonPtr = func();
    if (jsonPtr == nullptr) {
      throw RenSDKError('Failed to generate keypair');
    }

    final jsonStr = jsonPtr.toDartString();
    final freeFunc = _library
        .lookupFunction<
          Void Function(Pointer<Utf8>),
          void Function(Pointer<Utf8>)
        >('ren_sdk_free_string');
    freeFunc(jsonPtr);

    return Map<String, String>.from(jsonDecode(jsonStr));
  }

  /// Генерирует соль
  static String generateSalt() {
    final func = _library
        .lookupFunction<Pointer<Utf8> Function(), Pointer<Utf8> Function()>(
          'ren_sdk_generate_salt',
        );

    final saltPtr = func();
    if (saltPtr == nullptr) {
      throw RenSDKError('Failed to generate salt');
    }

    final salt = saltPtr.toDartString();
    final freeFunc = _library
        .lookupFunction<
          Void Function(Pointer<Utf8>),
          void Function(Pointer<Utf8>)
        >('ren_sdk_free_string');
    freeFunc(saltPtr);

    return salt;
  }
}

File: examples/ios/RenSDK.swift
//
//  RenSDK.swift
//  Ren Messenger iOS
//
//  Swift обёртка для Ren SDK (C FFI)
//

import Foundation

/// Ошибка SDK
public enum RenSDKError: Error {
    case invalidHandle
    case invalidParameters
    case apiError(String)
    case unknown
}

/// Клиент Ren SDK
public class RenSDK {
    private var handle: OpaquePointer?
    private let baseURL: String
    
    /// Инициализирует клиент SDK
    public init(baseURL: String) {
        self.baseURL = baseURL
        self.handle = ren_sdk_client_new(baseURL)
    }
    
    deinit {
        if let handle = handle {
            ren_sdk_client_free(handle)
        }
    }
    
    /// Устанавливает токен авторизации
    public func setToken(_ token: String) throws {
        guard let handle = handle else {
            throw RenSDKError.invalidHandle
        }
        
        let result = token.withCString { tokenPtr in
            ren_sdk_client_set_token(handle, tokenPtr)
        }
        
        if result.code != 0 {
            let message = result.message != nil ? String(cString: result.message!) : "Unknown error"
            ren_sdk_free_string(result.message)
            throw RenSDKError.apiError(message)
        }
    }
    
    /// Получает токен
    public func getToken() -> String? {
        guard let handle = handle else {
            return nil
        }
        
        guard let tokenPtr = ren_sdk_client_get_token(handle) else {
            return nil
        }
        
        let token = String(cString: tokenPtr)
        ren_sdk_free_string(tokenPtr)
        return token
    }
    
    /// Выполняет вход в систему
    public func login(login: String, password: String, rememberMe: Bool = false) throws {
        guard let handle = handle else {
            throw RenSDKError.invalidHandle
        }
        
        let result = login.withCString { loginPtr in
            password.withCString { passwordPtr in
                ren_sdk_login(handle, loginPtr, passwordPtr, rememberMe ? 1 : 0)
            }
        }
        
        if result.code != 0 {
            let message = result.message != nil ? String(cString: result.message!) : "Unknown error"
            ren_sdk_free_string(result.message)
            throw RenSDKError.apiError(message)
        }
    }
    
    /// Получает профиль текущего пользователя
    public func getMe() throws -> [String: Any] {
        guard let handle = handle else {
            throw RenSDKError.invalidHandle
        }
        
        guard let jsonPtr = ren_sdk_get_me(handle) else {
            throw RenSDKError.unknown
        }
        
        let jsonString = String(cString: jsonPtr)
        ren_sdk_free_string(jsonPtr)
        
        guard let data = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw RenSDKError.unknown
        }
        
        return json
    }
    
    /// Получает список чатов
    public func getChats() throws -> [[String: Any]] {
        guard let handle = handle else {
            throw RenSDKError.invalidHandle
        }
        
        guard let jsonPtr = ren_sdk_get_chats(handle) else {
            throw RenSDKError.unknown
        }
        
        let jsonString = String(cString: jsonPtr)
        ren_sdk_free_string(jsonPtr)
        
        guard let data = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw RenSDKError.unknown
        }
        
        return json
    }
    
    /// Генерирует пару ключей
    public static func generateKeypair() throws -> [String: String] {
        guard let jsonPtr = ren_sdk_generate_keypair() else {
            throw RenSDKError.unknown
        }
        
        let jsonString = String(cString: jsonPtr)
        ren_sdk_free_string(jsonPtr)
        
        guard let data = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: String] else {
            throw RenSDKError.unknown
        }
        
        return json
    }
    
    /// Генерирует соль
    public static func generateSalt() -> String {
        guard let saltPtr = ren_sdk_generate_salt() else {
            return ""
        }
        
        let salt = String(cString: saltPtr)
        ren_sdk_free_string(saltPtr)
        return salt
    }
}

// MARK: - C FFI функции

@_silgen_name("ren_sdk_client_new")
func ren_sdk_client_new(_ baseURL: UnsafePointer<CChar>) -> OpaquePointer?

@_silgen_name("ren_sdk_client_free")
func ren_sdk_client_free(_ handle: OpaquePointer)

@_silgen_name("ren_sdk_client_set_token")
func ren_sdk_client_set_token(_ handle: OpaquePointer, _ token: UnsafePointer<CChar>) -> RenResult

@_silgen_name("ren_sdk_client_get_token")
func ren_sdk_client_get_token(_ handle: OpaquePointer) -> UnsafeMutablePointer<CChar>?

@_silgen_name("ren_sdk_login")
func ren_sdk_login(_ handle: OpaquePointer, _ login: UnsafePointer<CChar>, _ password: UnsafePointer<CChar>, _ rememberMe: Int32) -> RenResult

@_silgen_name("ren_sdk_get_me")
func ren_sdk_get_me(_ handle: OpaquePointer) -> UnsafeMutablePointer<CChar>?

@_silgen_name("ren_sdk_get_chats")
func ren_sdk_get_chats(_ handle: OpaquePointer) -> UnsafeMutablePointer<CChar>?

@_silgen_name("ren_sdk_generate_keypair")
func ren_sdk_generate_keypair() -> UnsafeMutablePointer<CChar>?

@_silgen_name("ren_sdk_generate_salt")
func ren_sdk_generate_salt() -> UnsafeMutablePointer<CChar>?

@_silgen_name("ren_sdk_free_string")
func ren_sdk_free_string(_ str: UnsafeMutablePointer<CChar>)

struct RenResult {
    let code: Int32
    let message: UnsafeMutablePointer<CChar>?
}

